contracts/multisig/lib.rs

```diff
@@ -0,0 +1,303 @@
+  // of keys will they be present in this event.
+  #[ink(event)]
+  pub struct Vote {
+    /// Validator who issued the vote.
+    #[ink(topic)]
+    validator: AccountId,
+    /// Validator set for which keys are being generated.
+    #[ink(topic)]
+    validator_set: [u8; 32],
+    /// Hash of the keys voted on.
+    #[ink(topic)]
+    hash: [u8; 32],
+    /// Keys voted on.
+    keys: Option<Vec<Vec<u8>>>,
+  }
+
+  /// Event emitted when the new keys are fully generated for all curves, having been fully voted
+  /// on.
+  #[ink(event)]
+  pub struct KeyGen {
+    #[ink(topic)]
+    validator_set: [u8; 32],
+    #[ink(topic)]
+    hash: [u8; 32],
+  }
+
+  /// The Multisig error types.
+  #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+  #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+  pub enum Error {
+    /// Returned if a curve index doesn't have a key registered for it.
+    NonExistentCurve,
+    /// Returned if a non-validator is voting.
+    NotValidator,
+    /// Returned if this validator set already generated keys.
+    AlreadyGeneratedKeys,
+    /// Returned if this validator has already voted for these keys.
+    AlreadyVoted,
+  }
+
+  /// The Multisig result type.
+  pub type Result<T> = core::result::Result<T, Error>;
+
+  impl Multisig {
+    /// Deploys the Multisig contract.
+    #[ink(constructor)]
+    pub fn new() -> Self {
+      ink_lang::utils::initialize_contract(|_| {})
+    }
+
+    /// Validator set currently holding the multisig.
+    #[ink(message)]
+    pub fn validator_set(&self) -> [u8; 32] {
+      self.validator_set
+    }
+
+    /// Returns the key currently in-use for a given curve ID. This is then bound to a given chain
+    /// by applying a personalized additive offset, as done by the processor. Each chain then has
+    /// its own way of receiving funds to these keys, leaving this not for usage by wallets, nor
+    /// the processor which is expected to track events for this information. This is really solely
+    /// for debugging purposes.
+    #[ink(message)]
+    pub fn key(&self, curve: u8) -> Result<Vec<u8>> {
+      self.keys.get(curve).ok_or(Error::NonExistentCurve)
+    }
+
+    // TODO: voted
+    // TODO: votes
+
+    fn hash<T: Encode>(value: &T) -> [u8; 32] {
+      let mut output = [0; 32];
+      hash_encoded::<Blake2x256, _>(value, &mut output);
+      output
+    }
+
+    /// Vote for a given set of keys.
+    #[ink(message)]
+    pub fn vote(&mut self, keys: Vec<Vec<u8>>) -> Result<()> {
+      if keys.len() > 256 {
+        Err(Error::NonExistentCurve)?;
+      }
+
+      let validator = self.env().caller();
+      if !self.env().extension().is_active_validator(&validator) {
+        Err(Error::NotValidator)?;
+      }
+
+      let validator_set = self.env().extension().validator_set_id();
+      if self.validator_set == validator_set {
+        Err(Error::AlreadyGeneratedKeys)?;
+      }
```

Comment:
I'm a bit confused as to what this is checking :p

Reply:
It's checking that a validator set can't have multiple keys provided. If a keygen supposedly fails by one vote, and a new keygen succeeds, it shouldn't be possible for a single party to tell the chain the old one succeeded as well by submitting their vote AFTER having already performed a keygen.